### PR TITLE
feat[lang]: add `module.__at__()` to cast to interface

### PR DESCRIPTION
**Pull request by charles-cooper**
### What I did
fixes https://github.com/vyperlang/vyper/issues/3943, https://github.com/vyperlang/vyper/issues/3959

### Commit message

```
add `module.__at__`, a new `MemberFunctionT`, which allows the user to
cast addresses to a module's interface.

additionally, fix a bug where interfaces defined inline could not
be exported. this is simultaneously fixed as a related bug because
previously, interfaces could come up in export analysis as `InterfaceT`
or `TYPE_T` depending on their provenance. this commit fixes the bug by
making them `TYPE_T` in both imported and inlined provenance.

this also allows `module.__interface__` to be used in export position
by adding it to `ModuleT`'s members. note this has an unwanted side
effect of allowing `module.__interface__` in call position; in other
words, `module.__interface__(<address>)` has the same behavior as
`module.__at__(<address>)` when use as an expression. this can be
addressed in a later refactor.

refactor:
- wrap interfaces in `TYPE_T`
- streamline an `isinstance(t, (VyperType, TYPE_T))` check. TYPE_T` now
  inherits from `VyperType`, so it doesn't need to be listed separately
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/semantics/types/module.py [hunk 1]
```diff
@@ -323,16 +323,23 @@ def __init__(self, module: vy_ast.Module, name: Optional[str] = None):
...
 
             if hasattr(import_info.typ, "module_t"):
-                self._helper.add_member(import_info.alias, TYPE_T(import_info.typ))
+                module_info = import_info.typ
+                # get_expr_info uses ModuleInfo
+                self.add_member(import_info.alias, module_info)
+                # type_from_annotation uses TYPE_T
+                self._helper.add_member(import_info.alias, TYPE_T(module_info.module_t))
+            else:  # interfaces
+                self.add_member(import_info.alias, TYPE_T(import_info.typ))
```
Comment (cyberthirst) [inline vyper/semantics/types/module.py:334]: can we add an assert for it being an interface?
might be useful in the future should we allow eg selective imports

**Comment by cyberthirst** [comment]
we allow for exporting a module with no external functions - i think we should raise in such case:
```vyper
# main.vy
import lib1

def f(x: Bytes[32*5]) -> uint256:
     k: uint256 = lib1.foo()
     return k


exports: lib1.__interface__

# lib1.vy
@internal
def foo() -> uint256:
    return 1
```

**Comment by cyberthirst** [comment]
```vyper
import lib1

uses: lib1

@deploy
def __init__():
    lib1.__interface__(self).__init__()


exports: lib1.__interface__

#lib1.vy
k: uint256

@external
def bar():
    pass

@deploy
def __init__():
    self.k = 10
```

```
AttributeError: 'NoneType' object has no attribute 'module_t'
```

**Comment by cyberthirst** [comment]
`__interface__` can be used in "weird positions:
```vyper
main.vy
import lib1

#uses: lib1

@deploy
def __init__():
    log lib1.__interface__.Foo(1)
    s: lib1.Structt = lib1.__interface__.Structt(i=1)


exports: lib1.__interface__

#lib1.vy
event Foo:
    i: uint256

struct Structt:
    i: uint256

k: uint256

@external
def bar():
    pass

@deploy
def __init__():
    self.k = 10
```

**Comment by cyberthirst** [comment]
should address https://github.com/vyperlang/vyper/issues/3943, right?

File: vyper/semantics/types/module.py (hunk 1 above)
Comment (charles-cooper) [reply vyper/semantics/types/module.py:335]: 2d5f67df0e18e506918965a84722363b169c913e

**Comment by fubuloubu** [comment]
would it be too much to ask for implicit cast of `mod` to `mod.__interface__`? For either `v: Mod` in argument (convert to Interface) or `extcall Mod(v).method(...)`

**Comment by charles-cooper** [comment]
> with `-f annotated_ast` i get:
> 
> ```
> AttributeError: 'FunctionDef' object has no attribute 'node_id'
> ```
> 
> ```vyper
> # main.vy
> import lib1
> 
> @external
> def foo() -> uint256:
>     return staticcall lib1.__interface__(self).d()
> # lib1.vy
> d: public(uint256)
> ```

on latest (96213973ed07c584436ac9291b60061473649817) we get:
```
vyper.exceptions.UnknownAttribute: tmp/lib1.vy has no member 'd'.

  contract "tmp/main.vy:6", function "foo", line 6:22 
       5 def foo() -> uint256:
  ---> 6     return staticcall lib1.__interface__(self).d()
  -----------------------------^
       7
```

**Comment by charles-cooper** [comment]
> would it be too much to ask for implicit cast of `mod` to `mod.__interface__`? For either `v: Mod` in argument (convert to Interface to use like `v.method(...)`) or `extcall Mod(v).method(...)`
> 
> `.__interface__` is a little awkward to specify

yes but it's a bit hairy. it will feel ambiguous once we add `deploy Mod(...)` syntax

**Comment by charles-cooper** [comment]
> we allow for exporting a module with no external functions - i think we should raise in such case:
> 
> ```vyper
> # main.vy
> import lib1
> 
> def f(x: Bytes[32*5]) -> uint256:
>      k: uint256 = lib1.foo()
>      return k
> 
> 
> exports: lib1.__interface__
> 
> # lib1.vy
> @internal
> def foo() -> uint256:
>     return 1
> ```

24ac428539955f272a08e49301fbd8107d30b048

**Comment by charles-cooper** [comment]
> should address #3943, right?

yes. updated the PR description

**Comment by charles-cooper** [comment]
> with `-f annotated_ast` i get:
> 
> ```
> AttributeError: 'FunctionDef' object has no attribute 'node_id'
> ```
> 
> ```vyper
> # main.vy
> import lib1
> 
> @external
> def foo() -> uint256:
>     return staticcall lib1.__interface__(self).d()
> # lib1.vy
> d: public(uint256)
> ```

should be fixed as of latest (24ac428539955f272a08e49301fbd8107d30b048)

**Comment by charles-cooper** [comment]
> `__interface__` can be used in "weird positions:
> 
> ```vyper
> main.vy
> import lib1
> 
> #uses: lib1
> 
> @deploy
> def __init__():
>     log lib1.__interface__.Foo(1)
>     s: lib1.Structt = lib1.__interface__.Structt(i=1)
> 
> 
> exports: lib1.__interface__
> 
> #lib1.vy
> event Foo:
>     i: uint256
> 
> struct Structt:
>     i: uint256
> 
> k: uint256
> 
> @external
> def bar():
>     pass
> 
> @deploy
> def __init__():
>     self.k = 10
> ```

i'm not too bothered. it looks a bit weird, but technically `Structt` is a member of the interface. are you suggesting we should block it and only allow using `lib1.__interface__.<external function>`?

**Comment by cyberthirst** [comment]
a bit oos for this PR, but we allow for `implements: I` where `I` is an empty `.vyi` file - which most likely is a bug in the user code

**Comment by cyberthirst** [comment]
for code like
```vyper
# main.vy
import lib1

implements: lib1. __interface__
```
i get:
```
vyper.exceptions.UnknownAttribute: tests/custom/lib1.vy has no member '__interface__'.

  contract "tests/custom/test4.vy:4", line 4:12 
       3
  ---> 4 implements: lib1. __interface__
  -------------------^
       5
```

which is weird given we allow
```vyper
# main.vy
import lib1

exports: lib1. __interface__
```

**Comment by cyberthirst** [comment]
```vyper
# main.vy
import ITest2

exports: ITest2

# ITest2.vyi
def foo() -> uint8:
    ...
@view
def foobar() -> uint8:
    ...
 ```
 
yields:
```
Error compiling: tests/custom/test4.vy
vyper.exceptions.StructureException: not a function or interface: `type(tests/custom/ITest2.vyi)`

  contract "tests/custom/test4.vy:4", line 4:9 
       3
  ---> 4 exports: ITest2
  ----------------^
       5

````

**Comment by cyberthirst** [comment]
> i'm not too bothered. it looks a bit weird, but technically `Structt` is a member of the interface. are you suggesting we should block it and only allow using `lib1.__interface__.<external function>`?

yes, that was the original intention

however, as i test it, the poc now doesn't work. is this intentional?

**Comment by charles-cooper** [comment]
> for code like
> 
> ```vyper
> # main.vy
> import lib1
> 
> implements: lib1. __interface__
> ```
> 
> i get:
> 
> ```
> vyper.exceptions.UnknownAttribute: tests/custom/lib1.vy has no member '__interface__'.
> 
>   contract "tests/custom/test4.vy:4", line 4:12 
>        3
>   ---> 4 implements: lib1. __interface__
>   -------------------^
>        5
> ```
> 
> which is weird given we allow
> 
> ```vyper
> # main.vy
> import lib1
> 
> exports: lib1. __interface__
> ```

i can't reproduce this. can you provide the contents you have for `lib1.vy`?

**Comment by charles-cooper** [comment]
> ```vyper
> # main.vy
> import ITest2
> 
> exports: ITest2
> 
> # ITest2.vyi
> def foo() -> uint8:
>     ...
> @view
> def foobar() -> uint8:
>     ...
> ```
> 
> yields:
> 
> ```
> Error compiling: tests/custom/test4.vy
> vyper.exceptions.StructureException: not a function or interface: `type(tests/custom/ITest2.vyi)`
> 
>   contract "tests/custom/test4.vy:4", line 4:9 
>        3
>   ---> 4 exports: ITest2
>   ----------------^
>        5
> ```

i can't reproduce this. when i run this i get the error:
```
Error compiling: tmp/main.vy
vyper.exceptions.StructureException: invalid export

  contract "tmp/main.vy:6", line 6:0 
       5
  ---> 6 exports: ITest2
  -------^
       7

  (hint: exports should look like <module>.<function | interface>)
```

**Comment by charles-cooper** [comment]
> a bit oos for this PR, but we allow for `implements: I` where `I` is an empty `.vyi` file - which most likely is a bug in the user code

https://github.com/vyperlang/vyper/pull/4322

**Comment by cyberthirst** [comment]
exports of the following form are allowed:

```vyper
# main.vy
import lib2

a: address

exports: lib2.__interface__(self.a).__default__

# lib2.vy
@external
def __default__():
    pass

```

**Comment by charles-cooper** [comment]
> exports of the following form are allowed:
> 
> ```vyper
> # main.vy
> import lib2
> 
> a: address
> 
> exports: lib2.__interface__(self.a).__default__
> 
> # lib2.vy
> @external
> def __default__():
>     pass
> ```

ac43bebccc5a88e7a9b31e9ca5c5cbdc46f0fc5a

**Comment by charles-cooper** [comment]
> > would it be too much to ask for implicit cast of `mod` to `mod.__interface__`? For either `v: Mod` in argument (convert to Interface to use like `v.method(...)`) or `extcall Mod(v).method(...)`
> > `.__interface__` is a little awkward to specify
> 
> yes but it's a bit hairy. it will feel ambiguous once we add `deploy Mod(...)` syntax

been thinking about this more and maybe it's feasible? since we already guard external calls with `extcall` or `staticcall`, you can kind of tell what type you get from the call AST context
```vyper
x: Mod = deploy Mod(args...)
   ^^^          ^^^^
   type: interface
                type: module ctor

extcall Mod(...)
        ^^^
        type: interface
```

it will be kind of confusing to have `Mod` be a module or interface depending on context, though.

**Comment by cyberthirst** [comment]
```vyper
# lib1.vy
def doo():
    pass

# main.vy
import lib1

i: public(lib1.__interface__)

@external
def boo():
    pass

exports: self.i
```

```
Error compiling: tests/custom/test.vy
AttributeError: 'NoneType' object has no attribute '_metadata'

During handling of the above exception, another exception occurred:

vyper.exceptions.CompilerPanic: unhandled exception 'NoneType' object has no attribute '_metadata'

  contract "tests/custom/test.vy:9", line 9:0 
       8
  ---> 9 exports: self.i
  -------^


This is an unhandled internal compiler error. Please create an issue on Github to notify the developers!
https://github.com/vyperlang/vyper/issues/new?template=bug.md
```

**Review by cyberthirst — APPROVED** [review]

**Review by pcaversaccio — APPROVED** [review]

**Pull request merged**